사용자 로그인 기능 추가

dlrtn opened the pull request.

dlrtn committed:
feat: 사용자 로그인용 DTO 생성

dlrtn committed:
feat: 사용자 로그인 기능 생성

dlrtn committed:
fix: 사용자 로그인용 DTO 수정

입력된 패스워드, 검증 수준 약화

dlrtn committed:
feat: jwt 의존성 추가

dlrtn committed:
feat: jwt 모듈 생성

dlrtn committed:
feat: 사용자 로그인 기능 수정

- jwt access token 반환하도록 수정
- jwt 모듈 의존성 추가
- 로그인 로직 수정(입력 정보 어떤 것 틀렸는지 모르게)

dlrtn committed:
chore: swagger 모듈 수정

- 불필요한 EnvService 제공문 삭제

dlrtn committed:
feat: redis 의존성 추가

dlrtn committed:
feat: redis 모듈 생성

dlrtn committed:
feat: jwt 기능 수정

- AccessToken, RefreshToken 각 발급하도록 Jwt 수정

dlrtn committed:
feat: 사용자 로그인 기능 수정

- Redis 모듈 의존성 추가
- RefreshToken 생성해서 redis 서버에 저장하는 기능 추가

dlrtn committed:
feat: RedisService 수정으로 인한 변경사항 반영

- RedisService Default 문 삭제

dlrtn committed:
test: 사용자 도메인 단위 테스트 작성

- mock 사용하여 테스트 진행

DeadRiceCake commented on src/common/redis/redis.service.ts:
```diff
@@ -0,0 +1,42 @@
+import { EnvService } from '../env/env.service';
+import { Injectable } from '@nestjs/common';
+
+@Injectable()
+export class RedisService {
+  public client: RedisClientType<RedisModules>;
+
+  constructor(private readonly envService: EnvService) {
+    this.client = createClient({
+      url: this.envService.get('REDIS_URL'),
+    });
+
+    this.client.on('error', (error) => {
+      console.log('redis error: ', error);
+    });
+
+    this.client.connect();
+  }
+
+  public async set(key: string, value: string): Promise<void> {
+    await this.client.sendCommand([
+      'SET',
+      key,
+      value,
+      'EX',
+      this.envService.get('REDIS_TTL'),
+    ]);
+  }
+
+  public async get(key: string): Promise<string> {
+    const result = await this.client.sendCommand(['GET', key.toString()]);
+
+    // todo : redis에서 null이나 undefined를 반환할 경우에 대한 처리 (만료된 키로 접근하거나 존재하지 않는 키)
+    if (result === null || result === undefined) {
+      throw new Error('존재하지 않는 키입니다. 다시 시도해 주세요.');
+    }
+
+    return result.toString();
+  }
+}
```
레디스는 서비스 형태보단 도메인 방식으로 구조를 바꿔보는게 좋을 것 같네용

dlrtn replied:
랴오랴오츠.. 히 이즈 갓..

DeadRiceCake commented on src/business/member/api/member.controller.ts:
```diff
@@ -34,4 +35,16 @@ export class MemberController {
+  @Post('/login')
+  @ApiOperation({ summary: '사용자 로그인 API' })
+  @ApiOkResponse({ description: '사용자 인증을 통해 사용자 정보를 반환한다.' })
+  async login(@Res() res: Response, @Body() request: MemberLoginRequestDto) {
```
RequestDTO를 실제로 검증하기 위해 validationPipe를 적용해 줘야 할 것 같슴다

dlrtn replied:
적용.. 완룟..!

DeadRiceCake commented on src/business/member/application/member.service.ts:
```diff
@@ -21,4 +28,23 @@ export class MemberService {
+    const member = await this.memberRepository.findByEmail(request.email);
+
+    if (!member || member.getPassword() !== request.password) {
+      throw new Error('입력 정보가 부정확합니다.');
```
기본 Error 객체를 던지기 보단 BadRequestException 객체를 던지는게 좋을 듯 합니다

dlrtn replied:
이것도.. 완룟..

DeadRiceCake reviewed (commented):
화이또

dlrtn committed:
refactor: redis 사용 라이브러리 변경으로 인한 반영사항

- 기존 redis 접속 모듈 관련 코드 제거
- 신규 redis 연동 라이브러리 의존성 추가
- 사용자 서비스 로그인 로직 수정
- 사용자 서비스 테스트 코드 수정
    - 의존성 위주로 수정

dlrtn committed:
feat: DTO 필드 검증 기능 추가

- 검증 파이프 전역 옵션으로 추가
- 유저 회원가입 DTO 패스워드 검증 필드 관련 옵션 수정

dlrtn committed:
chore: Error -> Exception로 변경

- 비밀번호 불일치 시 기존 Error에서 BadRequestException으로 변경
- 만약 로그인 정보의 이메일이 DB에 존재하지 않을 경우 EntityNotFoundError를 뱉는데,, 이건 어떡하지..?
